### PR TITLE
Improved links finding

### LINKED ISSUE #64: Special characters need to be encoded
Try the following file:

``` md
- [pass](http://httpbin.org/status/200)
- [fail](http://httpbin.org/get?comma=,&breaks+things&0001)
- [pass](http://httpbin.org/get?comma=%2c&encoded+works&0002)
- [fail](http://httpbin.org/get?space= &breaks+things&0003)
- [pass](http://httpbin.org/status/500)
```

Outputs:

```
> Checking links in ab
Links to check: 5
  1. http://httpbin.org/status/200
  2. http://httpbin.org/get?comma=
  3. http://httpbin.org/get?comma=%2c&encoded+works&0002
  4. http://httpbin.org/get?space=
  5. http://httpbin.org/status/500
Checking URLs: ✓✓?✓✓

Issues :-(
> Links 
  1. 500 http://httpbin.org/status/500
> Dupes 
  None ✓
```

As you can see the second and fourth URLs were sliced where the comma and space exist.

This is perfectly working markdown so `awesome_bot` should not be requiring the characters to be encoded.

### PR DESCRIPTION
**Pull request by KishanBagaria**
Fixed #64

### CODE REVIEW

**Comment by dkhamsing** [comment]
@KishanBagaria thanks for the pr, unfortunately the change broke some tests.. will dig into it later but if you want to check em out

- https://circleci.com/gh/dkhamsing/awesome_bot/136
- https://travis-ci.org/dkhamsing/awesome_bot/builds/122910896

**Comment by KishanBagaria** [comment]
I was just checking that out, it doesn't work with nested links...

**Comment by KishanBagaria** [comment]
Updated. CircleCI build is failing (idk why) but TravisCI build is passing.

**Comment by dkhamsing** [comment]
@KishanBagaria Thanks!
If you have the code, can you run `rspec`?

**Comment by KishanBagaria** [comment]
I just noticed I added the new dependency in the `Gemfile` instead of the `gemspec`

**Comment by dkhamsing** [comment]
Ah ic

**Comment by dkhamsing** [comment]
Yea basically this broke some tests

**Comment by dkhamsing** [comment]
@KishanBagaria actually if you look at `travis ci`, you'd see link parsing is broken (finds no links)

**Comment by KishanBagaria** [comment]
A previous commit passed, then I added a space character in the regex to retry circleci, but that messed up the link parsing. I've pushed the previous commit again.

The rspec tests will pass if I convert every test's content to markdown (`http://link` to `<http://link>`), but do you think that's a good idea or should I add support for raw links?

**Comment by dkhamsing** [comment]
Should support raw links.. can it be done without affecting tests ?

**Comment by KishanBagaria** [comment]
What do you think of this:
- the user inputs file
- if the file has an extension and [`github-markup`](https://github.com/github/markup#markups) supports it, use `github-markup` to convert it to HTML and extract links from HTML using regex.
- else use your original implementation, which supports raw links etc.

Since awesome_bot is targeted at awesome lists, awesome lists is a GitHub thing, awesome lists will always be composed in a markup supported by GitHub.
So the `github-markup` gem will cover every case and the links will be extracted perfectly.

**Comment by dkhamsing** [comment]
Interesting.. seems kind of heavy for not a significant improvement

Will think about it


Thanks for your ideas

**Comment by KishanBagaria** [comment]
By default, you can just include two new dependencies: `github-markup` and `redcarpet` (for markdown).

If the user inputs an `asciidoc` file or something, they'll see a message like:
`do a "gem install asciidoctor" first`

**Comment by dkhamsing** [comment]
@KishanBagaria wouldn't it be easier to properly encode the url in `links_filter` 

a super simpler fix is just to replace `,` with `%2c`

```ruby
x.gsub(/'.*/, '').gsub(',','%2c').gsub(/,.*/, '').gsub('/:', '/')
```

**Comment by KishanBagaria** [comment]
That works great for me.

It may break in the future for someone in an edge case, so a long-term fix seemed better to me.

But this is lightweight and perfect for me.

**Comment by dkhamsing** [comment]
Thanks for reporting the issue, link parsing is not great but works for most cases

I will put this fix in shortly and maybe revisit in the future

File: spec/check_spec.rb [hunk 1]
```diff
@@ -4,19 +4,12 @@
   describe "check" do
     context "given 2 dupes" do
       it "returns 2" do
-        dupes = %(
-          https://github.com
-          https://github.com
-        )
+        dupes = %(https://github.com https://github.com)
```
Comment (dkhamsing) [inline spec/check_spec.rb]: @KishanBagaria can you explain this change?

Comment (KishanBagaria) [inline spec/check_spec.rb]: The markdown parser parses the leading whitespace as a code block:

          https://github.com
          https://github.com

https://github.com https://github.com

**Comment by KishanBagaria** [comment]
The tests pass but it doesn't support `asciidoc`, I just made the tests pass.

If you want to go the approach I mentioned earlier, I can fix this.

File: spec/check_spec.rb (hunk 1 above)
Comment (dkhamsing) [inline spec/check_spec.rb]: Ah
so this would work?

```
  dupes = %(https://github.com
https://github.com)%

Comment (KishanBagaria) [inline spec/check_spec.rb]: Yep

https://github.com
https://github.com

**Comment by dkhamsing** [comment]
There are some good ideas here, will require some more thoughts before continuing

Thanks

**Comment by dkhamsing** [comment]
Specifically, I wonder if we can improve link parsing without adding any dependencies

**Comment by KishanBagaria** [comment]
Without using a dependency, you've to implement specific functions to extract:
  - inline and footnote links in markdown files (should support nesting)
  - raw links from asciidoc and other markup formats (original implementation in this case)

If a single new dependency is okay:
- use kramdown (since it supports GFM) for markdown files
- for other formats
  - if user installed gem for format, use gem
  - else use original implementation

**Comment by dkhamsing** [comment]
Favoring using less dependencies at the moment with next version `1.8.0`, closing for now
Thank you!

**Pull request closed without merging**